src/lib/support/Pool.h

```diff
@@ -0,0 +1,125 @@
+/*
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *   Defines a memory pool class BitMapObjectPool.
+ */
+
+#pragma once
+
+#include <array>
+#include <assert.h>
+#include <atomic>
+#include <limits>
+#include <new>
+#include <stddef.h>
+
+namespace chip {
+
+class StaticAllocatorBase
+{
+public:
+    StaticAllocatorBase(size_t capacity) : mAllocated(0), mCapacity(capacity) {}
```

Comment:
Have we looked at how code size is impacted by comparing definition-in-header vs. definition-in-source? Historical experience indicated that in-header was a win for one instantiation; however, thereafter, in-source and method calls pulled ahead.

Reply:
https://godbolt.org/z/osf79v

For 2 instantiations, the inline version is still shorter.